File: CPP/explore/MergeSortedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <cstdlib>

struct ListNode
{
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
  ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
  {

    if (l1 == NULL)
      return l2;
    if (l2 == NULL)
      return l1;

    // Dummy.
    ListNode *head = NULL;

    if (l1->val < l2->val)
    {
      head = l1;
      l1 = l1->next;
    }
    else
    {
      head = l2;
      l2 = l2->next;
    }

    ListNode *currentNode = head;

    while (l1 != NULL && l2 != NULL)
    {
      if (l1->val < l2->val)
      {
        currentNode->next = l1;
        l1 = l1->next;
      }
      else
      {
        currentNode->next = l2;
        l2 = l2->next;
      }

      currentNode = currentNode->next;
    }

    // Add to rest of the tail.
    if (l1 != NULL)
      currentNode->next = l1;
    else
      currentNode->next = l2;

    return head;
  }
};

File: JS/explore/easy/MergeSortedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
	// NULL case
	if (l1 === null) return l2;
	if (l2 === null) return l1;

	let head = null;

	if (l1.val < l2.val) {
		head = l1;
		l1 = l1.next;
	} else {
		head = l2;
		l2 = l2.next;
	}

	let currentNode = head;

	while (l1 !== null && l2 !== null) {
		if (l1.val < l2.val) {
			currentNode.next = l1;
			l1 = l1.next;
		} else {
			currentNode.next = l2;
			l2 = l2.next;
		}

		currentNode = currentNode.next;
	}

	if (l1 === null) currentNode.next = l2;
	else currentNode.next = l1;

	return head;
};
